core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -755,31 +891,37 @@ class KeyWithDupInfo(val key: SerializableKey, val oldRow: TiRow, var oldHandle:
   }
 }
 
+// Note: This object's hashCode return a fixed value which suggests it is not better use with
+// hash map, hash set, or any other places that use hashCode. It is only used in reduceByKey.
 class ToBeCheckedRow(val row: TiRow,
                      val handleKey: KeyWithDupInfo,
                      val indexKeys: List[KeyWithDupInfo])
     extends Serializable {
-
-  //  val idxId: Long
-  // if handle is conflict or any key in indices key conflict with each other
-  // then two row equals
+  // When one of the following case happens, two ToBeCheckedRows are considered as same:
+  // 1. handle key is same if not null
+  // 2. at least one of index keys is same and index keys is not empty
+  // 3. handle key is null and index keys are empty, but row is same.
   override def equals(that: Any): Boolean =
     that match {
       case row: ToBeCheckedRow =>
-        if (row == this) return true
-        if (indexKeys.isEmpty && row.indexKeys.isEmpty) return false
-        val mayConflict = row.indexKeys.toSet
-        this.indexKeys.exists(mayConflict)
+        // 1. handle key is same
+        var handleConflict = false
+        if (handleKey != null && row.handleKey != null) {
+          handleConflict = handleKey.equals(row.handleKey)
+        }
+        // 2. index keys are empty, just return false
+        if (indexKeys.isEmpty && row.indexKeys.isEmpty) {
+          return handleConflict
+        }
+        // 3. check is there any index key same
+        this.indexKeys.exists(row.indexKeys.toSet) || handleConflict
```

Comment:
maybe change to `handleConflict || this.indexKeys.exists(row.indexKeys.toSet)` for better performance